haystack/document_stores/pinecone.py

```diff
@@ -0,0 +1,682 @@
+
+        # Pinecone index params
+        self.replicas = replicas
+        self.shards = shards
+
+        # initialize dictionary of index connections
+        self.pinecone_indexes: Dict[str, pinecone.Index] = {}
+        clean_index = self._sanitize_index_name(index)
+        if pinecone_index:
+            self.pinecone_indexes[clean_index] = pinecone_index
+        else:
+            self.pinecone_indexes[clean_index] = self._create_index_if_not_exist(
+                vector_dim=self.vector_dim,
+                index=clean_index,
+                metric_type=self.metric_type,
+                replicas=self.replicas,
+                shards=self.shards,
+            )
+
+        self.return_embedding = return_embedding
+        self.embedding_field = embedding_field
+
+        self.progress_bar = progress_bar
+
+        super().__init__(
+            url=sql_url, index=index, duplicate_documents=duplicate_documents  # no sanitation for SQL index name
+        )
+
+        self._validate_index_sync()
+
+    def _sanitize_index_name(self, index: Optional[str]) -> Optional[str]:
+        if index is None:
+            return None
+        elif "_" in index:
+            return index.replace("_", "-").lower()
+        else:
+            return index.lower()
+
+    def _create_index_if_not_exist(
+        self,
+        vector_dim: int,
+        index: Optional[str] = None,
+        metric_type: Optional[str] = "cosine",
+        replicas: Optional[int] = 1,
+        shards: Optional[int] = 1,
+    ):
+        """
+        Create a new index for storing documents in case if an
+        index with the name doesn't exist already.
+        """
+        index = index or self.index
+        index = self._sanitize_index_name(index)
+
+        # if index already loaded can skip
+        if index in self.pinecone_indexes.keys():
+            index_conn = self.pinecone_indexes[index]
+        else:
+            # search pinecone hosted indexes and create if it does not exist
+            if index not in pinecone.list_indexes():
+                pinecone.create_index(
+                    name=index, dimension=vector_dim, metric=metric_type, replicas=replicas, shards=shards
+                )
+            index_conn = pinecone.Index(index)
+
+        # get index statistics
+        stats = index_conn.describe_index_stats()
+        dims = stats["dimension"]
+        count = stats["namespaces"][""]["vector_count"] if stats["namespaces"].get("") else 0
+        logger.info(f"Index statistics: name: {index}, embedding dimensions: {dims}, record count: {count}")
+        # return index connection
+        return index_conn
+
+    def _convert_pinecone_result_to_document(self, result: dict, return_embedding: bool) -> Document:
+        """
+        Convert Pinecone result dict into haystack document object. This is more involved because
+        weaviate search result dict varies between get and query interfaces.
```

Comment:
I think we should remove mentions of Weaviate here.

Reply:
updated